org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/util/JRTUtil.java

```diff
@@ -174,11 +148,11 @@ public static JrtFileSystem getJrtSystem(File image, String release) throws IOEx
 	 * @throws IOException
 	 *             on any error
 	 */
-	public static FileSystem getJrtFileSystem(String path) throws IOException {
+	public static FileSystem getJrtFileSystem(Path path) throws IOException {
 		try {
-			FileSystem fs = JRT_FILE_SYSTEMS.computeIfAbsent(path, p -> {
+			FileSystem fs = JRT_FILE_SYSTEMS.computeIfAbsent(path.toRealPath(), p -> {
```

Comment:
This ensures that in all cases only one `FileSystem` exists for the same JRT-file, regardless of the difficulties the path contains (being relative or having `..` elements, symlinks, etc.).

Reply:
Actually this can also be used in the two other caches using Path keys.